server/src/main/java/io/deephaven/server/barrage/BarrageStreamGenerator.java

```diff
@@ -289,11 +287,17 @@ public SubView(final BarrageStreamGenerator generator,
                 try (WritableRowSet intersect = keyspaceViewport.intersect(generator.rowsIncluded.original)) {
                     addRowOffsets = generator.rowsIncluded.original.invert(intersect);
                 }
-            } else if (!generator.rowsAdded.original.equals(generator.rowsIncluded.original)) {
-                // there are scoped rows included in the chunks that need to be removed
-                addRowOffsets = generator.rowsIncluded.original.invert(generator.rowsAdded.original);
             } else {
-                addRowOffsets = RowSetFactory.flat(generator.rowsAdded.original.size());
+                if (isInitialSnapshot) {
```

Comment:
This part has been re-written, but the primary problem is that during a long-running growing subscription we can see the same row snapshotted multiple times (to satisfy viewport subs, e.g.) in snapshot messages.

To prevent this, each subscription now maintains a shrinking rowset called `growingRemainingViewport` that tracks which rows are still required by the client.  Another rowset `growingIncrementalViewport` exists for each snapshot cycle to designate exactly which rows need to be sent to the client this cycle.

We now treat the messages specially when they are snapshot messages.  We can count on the server to restrict `rowsIncluded` properly for snapshot messages and can blindly accept all rows included.